src/libsyntax/parse/parser.rs

```diff
@@ -5107,7 +5109,32 @@ impl<'a> Parser<'a> {
                     if has_parens {
                         self.expect(&token::CloseDelim(token::Paren))?;
                     }
-                    let poly_trait = PolyTraitRef::new(lifetime_defs, path, lo.to(self.prev_span));
+                    let poly_span = lo.to(self.prev_span);
+                    if is_negative {
+                        let mut err = self
+                            .struct_span_err(poly_span,
+                                             "negative trait bounds are not supported");
+                        let mut span_to_remove = poly_span;
+                        // Look back for `+` or `:` to suggest removal
+                        if let Ok(prev_source) = self.sess
+                            .source_map().span_to_prev_source(poly_span) {
+                                if let Some(spaces) = prev_source
+                                    .rsplitn(2, &['+', ':'] as &[_]).next() {
+                                        span_to_remove = poly_span.with_lo(
+                                            BytePos(poly_span.lo().0
+                                                    - 1 // for the `+` or `:`
+                                                    - spaces.len() as u32))
+                                    }
+                            }
```

Comment:
Instead of looking back in the `source_map`, wouldn't it make more sense to modify `parse_ty_param` to pass in the `:` span to `parse_generic_bounds_common` (so that you can remove it if necessary) and also keep track of the prior `+` [later in this method (`!self.eat_plus()`)](https://github.com/rust-lang/rust/pull/57364/files#diff-da9d34ca1d0f6beee2838cf02e07345cR5149)? That way you will use either one of those two spans to create the suggestion span. Something along the lines of:

```
last_plus_sp = None;
if self.eat_plus() {
    last_plus_sp = Some(self.last_span);
}

/* ... */

let span_to_remove = last_plus_sp.unwrap_or(colon_sp).until(poly_span);
```

Reply:
thank you for your mentoring! I was quite ready to give up on rustfixing this. I'll need some time to grok and implement your method, and making more edge cases.

Since this is the parser, iteration time degraded big time compared to modifying the AST validation pass. Do you have any tip for reducing edit->test time, e.g. `incremental` in config.toml? I promise not to waste the project's CI budget like that.

Reply:
Sadly, hacking on `parser.rs` is just inherently slow. I run `./x.py test src/test/ui --stage1 --bless`, and that takes a while. Learning to use the profiler can be worth it, but for the most part I just switch to working on something else/fetch tea while `rustc` does its thing :-|